openmeter/billing/service/invoice.go

```diff
@@ -37,3 +43,224 @@ func (s *Service) ListInvoices(ctx context.Context, input billing.ListInvoicesIn
 		return invoices, nil
 	})
 }
+
+func (s *Service) GetInvoiceByID(ctx context.Context, input billing.GetInvoiceByIdInput) (billingentity.Invoice, error) {
+	return entutils.TransactingRepo(ctx, s.adapter, func(ctx context.Context, txAdapter billing.Adapter) (billingentity.Invoice, error) {
+		invoice, err := txAdapter.GetInvoiceById(ctx, input)
+		if err != nil {
+			return billingentity.Invoice{}, err
+		}
+
+		if input.Expand.WorkflowApps {
+			resolvedApps, err := s.resolveApps(ctx, input.Invoice.Namespace, invoice.Workflow.AppReferences)
+			if err != nil {
+				return billingentity.Invoice{}, fmt.Errorf("error resolving apps for invoice [%s]: %w", invoice.ID, err)
+			}
+
+			invoice.Workflow.Apps = &billingentity.ProfileApps{
+				Tax:       resolvedApps.Tax.App,
+				Invoicing: resolvedApps.Invoicing.App,
+				Payment:   resolvedApps.Payment.App,
+			}
+		}
+
+		return invoice, nil
+	})
+}
+
+func (s *Service) CreateInvoiceAsOf(ctx context.Context, input billing.CreateInvoiceAsOfInput) ([]billingentity.Invoice, error) {
+	if err := input.Validate(); err != nil {
+		return nil, billing.ValidationError{
+			Err: err,
+		}
+	}
+
+	return TransactingRepoForGatheringInvoiceManipulation(
+		ctx,
+		s.adapter,
+		input.Customer,
+		func(ctx context.Context, txAdapter billing.Adapter) ([]billingentity.Invoice, error) {
+			// let's resolve the customer's settings
+			customerProfile, err := s.GetProfileWithCustomerOverride(ctx, billing.GetProfileWithCustomerOverrideInput{
+				Namespace:  input.Customer.Namespace,
+				CustomerID: input.Customer.ID,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("fetching customer profile: %w", err)
+			}
+
+			asof := lo.FromPtrOr(input.AsOf, clock.Now())
+
+			// let's gather the in-scope lines and validate it
+			inScopeLines, err := s.gatherInscopeLines(ctx, input, txAdapter, asof)
+			if err != nil {
+				return nil, err
+			}
+
+			sourceInvoiceIDs := lo.Uniq(lo.Map(inScopeLines, func(l billingentity.Line, _ int) string {
+				return l.InvoiceID
+			}))
+
+			if len(sourceInvoiceIDs) > 0 {
```

Comment:
The overall idea was that we should be able to generate an empty invoice and then add lines. However, that would require a slightly different API due to the currency, so I have changed things to support the main use-case of creating an invoice from gathering lines.

This can also be used to generate a new invoice using the following approach:
- Create pending manual lines
- Create invoice from them